src/control/client/types.go

```diff
@@ -233,26 +256,149 @@ func (rm ResultSmdMap) String() string {
 	return buf.String()
 }
 
-// ScmModules is an alias for protobuf ScmModule message slice representing
-// a number of SCM modules installed on a storage node.
+// ClientCtrlrMap is an alias for query results of NVMe controllers (and
+// any residing namespaces) on connected servers keyed on address.
+type ClientCtrlrMap map[string]pb_types.CtrlrResults
 
-// ControllerFactory is an interface providing capability to connect clients.
-type ControllerFactory interface {
-	create(string, *security.TransportConfig) (Control, error)
+func (ccm ClientCtrlrMap) String() string {
+	var buf bytes.Buffer
+	servers := make([]string, 0, len(ccm))
+
+	for server := range ccm {
+		servers = append(servers, server)
+	}
+	sort.Strings(servers)
+
+	for _, server := range servers {
+		fmt.Fprintf(&buf, "%s:\n%s\n", server, ccm[server])
+	}
+
+	return buf.String()
 }
 
-// controllerFactory as an implementation of ControllerFactory.
-type controllerFactory struct {
-	log logging.Logger
+// ClientMountMap is an alias for query results of SCM regions mounted
+// on connected servers keyed on address.
+type ClientMountMap map[string]pb_types.MountResults
+
+func (cmm ClientMountMap) String() string {
+	var buf bytes.Buffer
+	servers := make([]string, 0, len(cmm))
+
+	for server := range cmm {
+		servers = append(servers, server)
+	}
+	sort.Strings(servers)
+
+	for _, server := range servers {
+		fmt.Fprintf(&buf, "%s:\n%s\n", server, cmm[server])
+	}
+
+	return buf.String()
 }
 
-// create instantiates and connects a client to server at given address.
-func (c *controllerFactory) create(address string, cfg *security.TransportConfig) (Control, error) {
-	controller := &control{
-		log: c.log,
+// ScmScanResult represents the result of scanning for SCM
+// modules installed on a storage node and SCM namespaces.
+type ScmScanResult struct {
+	Modules    storage.ScmModules
+	Namespaces storage.ScmNamespaces
+	Err        error
+}
+
+func (result *ScmScanResult) String() string {
+	switch {
+	case result.Err != nil:
+		return fmt.Sprintf("Error: %s", result.Err)
+	case len(result.Namespaces) > 0:
+		return fmt.Sprintf("SCM Namespaces:\n%s\n", result.Namespaces)
+	default:
+		return fmt.Sprintf("SCM Modules:\n%s\n", result.Modules)
 	}
+}
 
-	err := controller.connect(address, cfg)
+// ScmScanMap maps ScmModuleScanResult structs to the addresses
+// of remote servers identified by an address string.
+type ScmScanResults map[string]*ScmScanResult
 
-	return controller, err
+func (results *ScmScanResults) String() string {
+	var buf bytes.Buffer
+	servers := make([]string, 0, len(*results))
+
+	for server := range *results {
+		servers = append(servers, server)
+	}
+	sort.Strings(servers)
+
+	for _, server := range servers {
+		fmt.Fprintf(&buf, "%s:\n%s\n", server, (*results)[server])
+	}
+
```

Comment:
Wondering if perhaps this logic could be shared somehow since it's being reused and seems to always concern strings.

Reply:
this could be done with something like
```
func sortKeys(m map[string]interface{}) []string {
⇥       keys := make([]string, 0, len(m))

⇥       for key := range m {
⇥       ⇥       keys = append(keys, key)
⇥       }
⇥       sort.Strings(keys)

⇥       return keys
}
```
however I've floated this kind of idea in the past and @mjmac is particularly against using interfaces in this way (and I can appreciate his perspective given most things can be done in a different way). so the trade-off for not using interfaces to handle generics is that you get this kind of code duplication. Also because we have custom type aliases for the maps this might require some casting.

Reply:
Understood. I think the generic interface actually makes a lot of sense here since we're just harvesting keys of a specific type from a map and don't care about the values. But I'm okay with leaving it as-is, the duplication just stood out to me.